src/set/mod.rs

```diff
@@ -0,0 +1,227 @@
+//! Serialization modules for serializing lists of values as sets
+//!
+//! When using the serializers in these submodules, it may be beneficial to
+//! annotate the field with `#[serde(default)]` and
+//! `#[serde(skip_serializing_if = "<empty check>")]`. This will make sure
+//! that the field is omitted when empty. DynamoDB will return an error if
+//! an empty set is used.
+//!
+//! These serializers do not check for duplicate values. If the set contains
+//! duplicate values, DynamoDB will return a validation error when the
+//! set is used.
+//!
+//! # Examples
+//!
+//! ```
+//! use std::collections::{HashSet, VecDeque};
+//!
+//! use serde_bytes::ByteBuf;
+//! use serde_derive::{Serialize, Deserialize};
+//! use serde_dynamo::{Item, AttributeValue};
+//!
+//! #[derive(Serialize, Deserialize)]
+//! struct MyStruct {
+//!     #[serde(with = "serde_dynamo::set::strings")]
+//!     #[serde(default, skip_serializing_if = "Vec::is_empty")]
+//!     names: Vec<String>,
+//!     #[serde(with = "serde_dynamo::set::numbers")]
+//!     #[serde(default, skip_serializing_if = "HashSet::is_empty")]
+//!     numbers: HashSet<u64>,
+//!     #[serde(with = "serde_dynamo::set::bytes")]
+//!     #[serde(default, skip_serializing_if = "VecDeque::is_empty")]
+//!     data: VecDeque<ByteBuf>,
+//! }
+//!
+//! let my_struct = MyStruct {
+//!     names: vec!["John".to_string(), "Jane".to_string()],
+//!     numbers: [14, 25, 32].into_iter().collect(),
+//!     data: vec![ByteBuf::from(b"hello".to_vec()), ByteBuf::from(b"world".to_vec())].into(),
+//! };
+//!
+//! let serialized: Item = serde_dynamo::to_item(&my_struct).unwrap();
+//! assert_eq!(
+//!     serialized["names"],
+//!     AttributeValue::Ss(vec!["John".to_string(), "Jane".to_string()])
+//! );
+//! if let AttributeValue::Ns(mut sorted_numbers) = serialized["numbers"].clone() {
+//!     sorted_numbers.sort();
+//!     assert_eq!(
+//!         sorted_numbers,
+//!         vec!["14".to_string(), "25".to_string(), "32".to_string()]
+//!     );
+//! } else {
+//!     panic!("Expected numbers to be a set of numbers");
+//! }
+//! assert_eq!(
+//!     serialized["data"],
+//!     AttributeValue::Bs(vec![b"hello".to_vec(), b"world".to_vec()])
+//! );
+//! ```
+
+pub mod bytes;
+pub mod numbers;
+pub mod strings;
+
+/// Serializes the wrapped value as a string set
+///
+/// This is useful for [`to_attribute_value`][crate::to_attribute_value]
+/// when you want to serialize a sequence as a set of strings.
+///
+/// # Examples
+///
+/// ```
+/// use serde_dynamo::{set::StringSet, AttributeValue};
+///
+/// let set = vec![
+///     "orange",
+///     "apple",
+/// ];
+///
+/// let val: AttributeValue = serde_dynamo::to_attribute_value(StringSet(set)).unwrap();
+/// assert_eq!(val, AttributeValue::Ss(vec![
+///     "orange".to_string(),
+///     "apple".to_string(),
+/// ]));
+/// ```
+pub struct StringSet<T>(pub T);
```

Comment:
Should this have a `deserialize`, too?

Reply:
It intentionally does not, in order to lead consumers to use `#[serde(with)]` when working with fields in a larger type. This particular struct should only be used with `to_attribute_value`, and isn't necessary when doing `from_attribute_value` due to the support for lists and sets in the default deserializer.